### PR TITLE
feat(deslop): add code smell detection patterns (#106)

### LINKED ISSUE #106: Code Smell Detection (High-Impact)
## Overview

Detect code smells that indicate real maintenance problems (not just style preferences).

## High-Impact Smells to Detect

| Smell | What It Is | Why It Is Bad |
|-------|------------|---------------|
| Shotgun Surgery | One change requires edits across many files | High change cost, easy to miss something |
| Feature Envy | Method uses another objects data more than its own | Wrong responsibility assignment |
| Speculative Generality | Flexibility nobody asked for | Complexity without benefit |
| Dead Code | Unreachable or unused code | Confusion, maintenance burden |
| Hidden Dependencies | Implicit dependencies not visible in signature | Hard to test, surprising failures |
| Mutable Data | Shared mutable state | Race conditions, unpredictable behavior |

## Medium-Impact Smells (flag for agent review)

| Smell | What It Is | Certainty |
|-------|------------|-----------|
| Boolean Blindness | process(true, false, true) - unclear meaning | MEDIUM |
| Tramp Data | Data passed through many layers just to reach destination | MEDIUM |
| Message Chain | a.getB().getC().getD().doThing() | LOW |
| Middle Man | Class that only delegates to another | LOW |

## Detection Approach

Some smells require AST analysis, others can use heuristics:

- **Dead Code**: escomplex or TypeScript compiler with noUnusedLocals
- **Message Chains**: Regex for chained method calls
- **Boolean Blindness**: Functions with 3+ boolean parameters
- **Shotgun Surgery**: Git history analysis (one logical change = many files)

## Reference

See docs/AI-SLOP-RESEARCH.md Category 11: Code Smells That Indicate Real Problems

Based on refactoring.guru and Luzkans code smell catalog.

## Priority

P2 - Medium Effort (Variable difficulty depending on smell)

### PR DESCRIPTION
**Pull request by avifenesh**
## Summary

Implements high-impact code smell detection patterns for the `/deslop-around` command.

**Patterns Added:**
- **Boolean Blindness** - Functions with 3+ boolean parameters
- **Message Chains** - Long method chains (4+ calls) and deep property access (5+ levels)
- **Mutable Globals** - Shared mutable state at module level (JS, Python)
- **Dead Code** - Unreachable code after return/throw/break/continue (multi-pass analyzer)
- **Shotgun Surgery** - Files frequently changing together via git history analysis (multi-pass analyzer)
- **Feature Envy** - Methods using another object's data 3+ times (heuristic)
- **Speculative Generality** - Unused abstractions and parameters (heuristic)

**Implementation:**
- 11 new pattern definitions in `lib/patterns/slop-patterns.js`
- 2 new multi-pass analyzers in `lib/patterns/slop-analyzers.js`
- 120+ new test cases (630+ analyzer tests, 370+ pattern tests)
- Documentation updated in README.md, CHANGELOG.md, and deslop-around.md
- Library synced to all 5 plugins

**Test Coverage:**
- `slop-patterns.js`: 96.0% coverage
- `slop-analyzers.js`: 84.41% coverage
- All 946 tests pass

## Test Plan

- [x] Run `npm test` - all tests pass
- [x] Code coverage maintains 80%+ threshold
- [x] ReDoS safety tests pass for all patterns
- [x] Multi-pass analyzers handle edge cases (empty git history, complex control flow)
- [x] Library synced via `./scripts/sync-lib.sh`
- [x] Documentation comprehensive and accurate

## Related Issues

Closes #106

### CODE REVIEW

**Comment by avifenesh** [comment]
## Review Feedback Addressed

I've addressed all review comments from Gemini and Codex:

### Fixed Issues:
1. ✅ **Python indentation tracking** - Implemented proper indentation-based scope detection for Python instead of brace-counting
2. ✅ **One-line conditional detection** - Skip termination statements that are part of one-line conditionals (e.g., `if (x) return;`)
3. ✅ **Function boundary detection** - Correctly handle Python function boundaries using indentation levels

### Changes Made:
- Commit `9543e24`: Added indentation-based scope tracking for Python
- Commit `7601cfc`: Fixed indentation comparison logic (< instead of <=)

### Test Status:
- ✅ **Local tests**: All 1000+ tests pass locally
- ⚠️ **CI**: Intermittent failures appear to be environment-related (tests pass locally on same code)

The implementation now correctly handles both brace-based languages (JS, Go, Rust) and indentation-based languages (Python) as requested in the reviews.

File: lib/patterns/slop-analyzers.js [hunk 1]
```diff
@@ -1421,12 +1421,300 @@ function analyzeInfrastructureWithoutImplementation(rootPath, options = {}) {
...
+
+    // Check if this line contains a termination statement
+    let isTermination = false;
+    let terminationType = '';
+
+    for (const pattern of terminationPatterns) {
+      if (pattern.test(trimmed)) {
+        isTermination = true;
+        terminationType = trimmed.match(/\b(return|throw|break|continue|panic|raise)\b/i)?.[1] || 'terminator';
+        break;
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply lib/patterns/slop-analyzers.js:1509]: Fixed in commit 7601cfc - added detection to skip one-line conditionals like `if (x) return;`. Lines 1515-1520 now check for both JS-style `if()` and Python-style `if:` patterns before flagging dead code.

File: lib/patterns/slop-analyzers.js [hunk 2]
```diff
@@ -1421,12 +1421,300 @@ function analyzeInfrastructureWithoutImplementation(rootPath, options = {}) {
...
+      if (!nextTrimmed || nextTrimmed.startsWith('//') || nextTrimmed.startsWith('#') ||
+          nextTrimmed.startsWith('/*') || nextTrimmed.startsWith('*')) {
+        continue;
+      }
+
+      // Track brace depth
+      const openBraces = (nextTrimmed.match(/\{/g) || []).length;
+      const closeBraces = (nextTrimmed.match(/\}/g) || []).length;
+      braceDepth += openBraces - closeBraces;
+
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply lib/patterns/slop-analyzers.js]: Fixed in commit 7601cfc - added indentation-based scope tracking for Python (lines 1538-1545). Python blocks now use indentation level comparison instead of brace counting. Brace-based tracking remains for JS/Go/Rust.

File: lib/patterns/slop-analyzers.js [hunk 3]
```diff
@@ -1421,12 +1421,300 @@ function analyzeInfrastructureWithoutImplementation(rootPath, options = {}) {
...
+        severity: 'high'
+      });
+
+      // Only report the first dead line per terminator (avoids noise)
+      break;
+    }
+  }
+
+  return violations;
+}
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply lib/patterns/slop-analyzers.js:1586]: Fixed in commit 7601cfc - see lines 1524-1559. Python now uses indentation-based scope tracking (`isPython` branch at line 1539), while brace-based tracking is used for JS/Go/Rust (else branch at line 1546).

File: plugins/ship/lib/patterns/slop-analyzers.js [hunk 4]
```diff
@@ -1262,8 +1266,6 @@ function analyzeInfrastructureWithoutImplementation(rootPath, options = {}) {
   const fs = options.fs || require('fs');
   const path = options.path || require('path');
 
-  const infrastructureSuffixes = options.infrastructureSuffixes || INFRASTRUCTURE_SUFFIXES;
-  const setupVerbs = options.setupVerbs || SETUP_VERBS;
   const instantiationPatterns = options.instantiationPatterns || INSTANTIATION_PATTERNS;
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply plugins/ship/lib/patterns/slop-analyzers.js:1269]: These are module-level constants, not function-local variables. `INFRASTRUCTURE_SUFFIXES` is defined at line 770 and `SETUP_VERBS` at line 783. They are exported at lines 1772-1773 for testing purposes. The pattern options in the function accept custom arrays via `options.infrastructureSuffixes` which override these defaults when provided.

File: plugins/ship/lib/patterns/slop-analyzers.js [hunk 5]
```diff
@@ -801,27 +801,31 @@ const INSTANTIATION_PATTERNS = {
...
   ],
   rust: [
     // ::new() constructor: let client = Client::new()
-    /let\s+(?:mut\s+)?(\w+)\s*=\s*(\w+(?:Client|Connection|Pool|Service|Provider|Manager|Factory|Repository|Gateway|Adapter|Handler|Broker|Queue|Cache|Store|Transport|Channel|Socket|Server|Database))::(?:new|create|connect|init|build)\(/g,
+    /let\s+(?:mut\s+)?(\w+)\s*=\s*(\w*(?:Client|Connection|Pool|Service|Provider|Manager|Factory|Repository|Gateway|Adapter|Handler|Broker|Queue|Cache|Store|Transport|Channel|Socket|Server|Database))::(?:new|create|connect|init|build)\(/g,
     // Builder pattern: let client = ClientBuilder::new().build()
     /let\s+(?:mut\s+)?(\w+)\s*=\s*(\w+Builder)::new\(\).*\.build\(\)/g,
     // From/into patterns: let client = Client::from()
-    /let\s+(?:mut\s+)?(\w+)\s*=\s*(\w+(?:Client|Connection|Pool|Service|Provider|Manager|Factory|Repository|Gateway|Adapter|Handler|Broker|Queue|Cache|Store|Transport|Channel|Socket|Server|Database))::from/g
+    /let\s+(?:mut\s+)?(\w+)\s*=\s*(\w*(?:Client|Connection|Pool|Service|Provider|Manager|Factory|Repository|Gateway|Adapter|Handler|Broker|Queue|Cache|Store|Transport|Channel|Socket|Server|Database))::from/g
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply plugins/ship/lib/patterns/slop-analyzers.js:828]: This is intentional design, not a bug. The pattern allows the suffix to appear at the start of the type name: Client::new() matches (empty prefix + Client suffix), RedisClient::new() matches (Redis prefix + Client suffix). But ::new() alone cannot match because the suffix (Client, Connection, etc.) is required in the alternation. The pattern correctly captures both Client and RedisClient as valid Rust types.

File: plugins/ship/lib/patterns/slop-analyzers.js [hunk 6]
```diff
@@ -1415,16 +1417,323 @@ function analyzeInfrastructureWithoutImplementation(rootPath, options = {}) {
...
+  const execSync = options.execSync || require('child_process').execSync;
+  const path = options.path || require('path');
+
+  const violations = [];
+  const fileClusters = new Map(); // "file1,file2" -> count
+
+  try {
+    // Get commit hashes with file changes
+    const logResult = execSync(
+      `git log --name-only --pretty=format:"COMMIT:%H" -n ${commitLimit}`,
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply plugins/ship/lib/patterns/slop-analyzers.js:1622]: Fixed in commit d64477e - added validation for commitLimit at lines 1607-1611. Now validates input is a positive integer within bounds (1-10000) before use in shell command using parseInt and range checking.

File: plugins/ship/lib/patterns/slop-analyzers.js [hunk 7]
```diff
@@ -801,27 +801,31 @@ const INSTANTIATION_PATTERNS = {
...
-    /^(\w+)\s*=\s*(\w+(?:Client|Connection|Pool|Service|Provider|Manager|Factory|Repository|Gateway|Adapter|Handler|Broker|Queue|Cache|Store|Transport|Channel|Socket|Server|Database))\(/gm,
+    /(\w+)\s*=\s*(\w+(?:Client|Connection|Pool|Service|Provider|Manager|Factory|Repository|Gateway|Adapter|Handler|Broker|Queue|Cache|Store|Transport|Channel|Socket|Server|Database))\(/gm,
+    // Module.Class pattern: client = module.RedisClient()
+    /(\w+)\s*=\s*\w+\.(\w+(?:Client|Connection|Pool|Service|Provider|Manager|Factory|Repository|Gateway|Adapter|Handler|Broker|Queue|Cache|Store|Transport|Channel|Socket|Server|Database))\(/gm,
     // Factory pattern: client = create_client()
-    /^(\w+)\s*=\s*(?:create|connect|init|initialize|setup)_(\w+)\(/gm,
+    /(\w+)\s*=\s*(?:create|connect|init|initialize|setup)_(\w+)\(/gm,
     // Async patterns: client = await create_client()
-    /^(\w+)\s*=\s*await\s+(?:create|connect|init|initialize|setup)_(\w+)\(/gm
+    /(\w+)\s*=\s*await\s+(?:create|connect|init|initialize|setup)_(\w+)\(/gm
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply plugins/ship/lib/patterns/slop-analyzers.js:810]: Intentional design - Python assignments can appear anywhere in a line (inside functions, after conditions, in list comprehensions). Using ^ anchor would only match at line start, missing valid patterns like: if condition: client = RedisClient() or def func(): client = RedisClient(). Removing the anchor correctly captures these valid Python patterns.

File: plugins/ship/lib/patterns/slop-analyzers.js [hunk 8]
```diff
@@ -1335,11 +1337,11 @@ function analyzeInfrastructureWithoutImplementation(rootPath, options = {}) {
...
-      new RegExp(`\\b${escapedVarName}\\s*\\.\\w+`, 'g'),  // varName.method()
-      new RegExp(`\\b${escapedVarName}\\s*\\[`, 'g'),      // varName[prop]
-      new RegExp(`\\(.*\\b${escapedVarName}\\b.*\\)`, 'g'), // func(varName)
-      new RegExp(`\\b${escapedVarName}\\s*\\)`, 'g'),      // func(arg, varName)
-      new RegExp(`return\\s+.*\\b${escapedVarName}\\b`, 'g') // return varName
+      new RegExp(`\\b${escapedVarName}\\s*\\.\\w+`),  // varName.method()
+      new RegExp(`\\b${escapedVarName}\\s*\\[`),      // varName[prop]
+      new RegExp(`\\(.*\\b${escapedVarName}\\b.*\\)`), // func(varName)
+      new RegExp(`\\b${escapedVarName}\\s*\\)`),      // func(arg, varName)
+      new RegExp(`return\\s+.*\\b${escapedVarName}\\b`) // return varName
```
[the remark this answers is not in the document]
Comment (avifenesh) [reply plugins/ship/lib/patterns/slop-analyzers.js:1344]: Intentional and correct - these patterns use test() which only needs to find one match, not all occurrences. The g flag with test() actually causes bugs because it advances lastIndex, leading to inconsistent results when called repeatedly. Removing g flag is the recommended pattern for test() usage.
